### PR TITLE
feat(besu): Support for private transactions

### LINKED ISSUE #952: feat(besu): Add private transaction support
**Is your feature request related to a problem? Please describe.**

Feature request: I need to be able to issue private transactions to a Besu ledger via the Cactus API client.

**Describe the solution you'd like**

1. Either a separate endpoint next to the current POST /run-transaction one or as an extension of it. Not yet sure which one is more feasible/makes more sense.
2. Test coverage for the new feature
3. Private transactions must not be logged by Cactus for obvious reasons.
4. Documentation to explain specifically that this is not a privacy preserving feature, it only allows access to the pre-existing Besu APIs as-is

**Describe alternatives you've considered**

Nothing. There is no doubt in my mind that we need this feature.

Additional context

Here's the Besu documentation regarding private transactions:
https://besu.hyperledger.org/en/stable/Concepts/Privacy/Private-Transactions/

cc: @takeutak @sfuji822 @jonathan-m-hamilton @petermetz @kikoncuo @travis-payne

### PR DESCRIPTION
**Pull request by travis-payne**
Support for besu private transactions is now enabled, this is done by adding privateFor and privateFrom fields in the HTTP request.

fixes #952

### CODE REVIEW

**Review by takeutak — APPROVED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@travis-payne Please delete `packages/cactus-cockpit/package-lock.json` (we do not have/need package-lock files anymore since the migration to Yarn)

**Review by petermetz — CHANGES_REQUESTED** [review]
@travis-payne Seems like `examples/discounted-cartrade/README.md` has changes due to some merge/rebase gone wrong?

**Comment by petermetz** [comment]
@travis-payne Looks like maybe something has gone south with the rebase? Let me know if you need help fixing.

File: examples/discounted-cartrade/README.md [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # Cactus discounted-cartrade
```
Comment (petermetz) [inline examples/discounted-cartrade/README.md:2]: @travis-payne Nit: change might be by accident?

File: packages/cactus-plugin-ledger-connector-besu/README.md [hunk 2]
```diff
@@ -97,9 +97,13 @@ enum Web3SigningCredentialType {
     NONE = 'NONE'
 }
 ```
+
 > Extensive documentation and examples in the [readthedocs](https://readthedocs.org/projects/hyperledger-cactus/) (WIP) 
 
 
+## Transaction Privacy Feature
+Private transactions using Besu are currently enabled.
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-besu/README.md:105]: @travis-payne Please open a follow-up issue in GH to document their usage!

File: packages/cactus-test-cmd-api-server/package.json [hunk 3]
```diff
@@ -69,7 +69,8 @@
     "@hyperledger/cactus-core-api": "0.7.0",
     "@hyperledger/cactus-plugin-keychain-vault": "0.7.0",
     "express-jwt-authz": "2.4.1",
-    "jose": "1.28.1"
+    "jose": "1.28.1",
+    "artillery": "1.7.2"
```
Comment (petermetz) [inline packages/cactus-test-cmd-api-server/package.json:73]: @travis-payne The artillery benchmark is in the other package (not the test one but the cmd-api-server package itself).
Please either move this devDependency over there or bring the test case over here just so that they are in the same package.

`/packages/cactus-cmd-api-server/src/test/typescript/benchmark/artillery-api-benchmark.test.ts`

File: packages/cactus-test-tooling/src/main/typescript/besu/besu-test-ledger.ts [hunk 4]
```diff
@@ -145,7 +145,7 @@ export class BesuTestLedger implements ITestLedger {
    * @see https://github.com/hyperledger/besu/blob/1.5.1/config/src/main/resources/dev.json
    */
   public getGenesisAccountPubKey(): string {
-    return "627306090abaB3A6e1400e9345bC60c78a8BEf57";
+    return "0x627306090abaB3A6e1400e9345bC60c78a8BEf57";
```
Comment (petermetz) [inline packages/cactus-test-tooling/src/main/typescript/besu/besu-test-ledger.ts:148]: @travis-payne If not already please include in the commit message the research/explanation that we came up with for why this change was necesary (the whole big number investigation we did a while back, there was some GH issue for it in the web3 repo IIRC - a link to that would be great as well)

Sorry if all of this is already in the commit message I didn't check!

File: typings/web3-eea/index.d.ts [hunk 5]
```diff
@@ -1,5 +1,7 @@
-
 declare module "web3-eea" {
+    import type Web3 from "web3";
```
Comment (petermetz) [inline typings/web3-eea/index.d.ts:2]: @travis-payne Please create a follow-up task to remove this file and it's parent directory completely from our repo once this other PR gets merged (+released with the then latest version of web3-eea)
https://github.com/ConsenSys/web3js-eea/pull/166
(once the above happens this file will become redundant which is great because we don't want to maintain it ourselves within Cactus)

File: packages/cactus-plugin-ledger-connector-besu/src/main/typescript/plugin-ledger-connector-besu.ts [hunk 6]
```diff
@@ -389,8 +399,55 @@ export class PluginLedgerConnectorBesu
     const method: ContractSendMethod = methodRef(...req.params);
 
     if (req.invocationType === EthContractInvocationType.Call) {
-      const callOutput = await (method as any).call();
-      const success = true;
+      let callOutput;
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-besu/src/main/typescript/plugin-ledger-connector-besu.ts:402]: @travis-payne Please open a follow-up issue to come up with a way of simplyfing the code here, the length of the method and the number of if conditions starts to smell here. No need to re-work anything in the context of this current PR because this one really needs to move along now, but I want to make sure we track it that this has to be improved.

@TonyRowntree is already working on some other PRs that make the keychain optional by splitting the endpoints to smaller ones so we have to also cross-coordinate with him/his PRs. There will be conflicts, but we have to get this code complexity under control before the 1.0 release because then all  breaking changes will have to wait until Cactus 2.0.

cc: @m-courtin

File: packages/cactus-plugin-ledger-connector-besu/src/main/json/openapi.json [hunk 7]
```diff
@@ -641,6 +641,26 @@
                     }
                 }
             },
+            "BesuPrivateTransactionConfig" : {
```
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-besu/src/main/json/openapi.json:644]: @travis-payne Please open a follow-up issue to have the code support the other possible parameters as well (privacyGroupId and privateKey)

See the index.d.ts file for reference, quoting from there:

```typescript
 export interface ISendRawTransactionOptions extends ISendGenericTransactionOptions, TransactionConfig {
        /**
         * Private Key used to sign transaction with
         */
        readonly privateKey: string;
        /**
         * Enclave public key
         */
        readonly privateFrom: string;
        /**
         * Enclave keys to send the transaction to
         */
        readonly privateFor: string[];
        /**
         * Enclave id representing the receivers of the transaction
         */
        readonly privacyGroupId?: string;
        /**
         * (Optional) : If not provided, will be calculated using`eea_getTransctionCount`
         */
        readonly nonce?: number;
        /**
         * The address to send the transaction
         */
        readonly to?: string;
        /**
         * Data to be sent in the transaction
         */
        readonly data: string;
    }
```

**Review by petermetz — CHANGES_REQUESTED** [review]
Almost good to go. Had a few nits and some admin chores, but otherwise looking pretty solid!

File: packages/cactus-plugin-ledger-connector-besu/src/main/json/openapi.json (hunk 7 above)
Comment (travis-payne) [reply packages/cactus-plugin-ledger-connector-besu/src/main/json/openapi.json:644]: I think private key isn't needed, users add that as part of their Web3SigningCredential, but privacyGroupId is defo needed.

**Review by petermetz — APPROVED** [review]

**Review by petermetz — APPROVED** [review]
LGTM

**Comment by petermetz** [comment]
@travis-payne Easiest way to resolve yarn.lock conflicts:
1. delete the lock file
2. `yarn` on the terminal
3. lock file is now in a state that's good, commit it

Please make sure to squash everything into a single commit before merging.
